feat(rca): items management api

kdelemme opened the pull request:
Resolves https://github.com/elastic/kibana/issues/190323
Resolves https://github.com/elastic/kibana/issues/190332

## Summary

This PR introduces a basic esql investigation item type and a set of API for adding, listing and deleting investigation items.
We are not using it from the client yet, I'm planning on integrating with this in a follow up PR: and this is were the complexity lives.

I've also 🧹 cleaned up the shared plugin, so we have the rest_specs and the schemas separated. 

As you can see below, the esql item type requires a esql query (easy) and a suggestion (not easy, generated and coming from lens). Typing this will be tricky. I suggest we don't try.


## Testing

I would use the API to create an investigation, add some items like the following one, and then list the items and delete it:

```
curl --request POST \
  --url http://localhost:5601/kibana/api/observability/investigations/my-awesome-id/items \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'elastic-api-version: 2023-10-31' \
  --header 'kbn-xsrf: oui' \
  --data '{
	"type": "esql",
	"title": "My super chart",
	"params": {
		"esql": "from kbn-data-forge-fake_stack.admin-console-* | STATS AVG(http.response.bytes) by event.action",
		"suggestion": {
			"id": "6ac3e467-ae63-46d2-b54a-91fffc50daba",
			"title": "Donut",
			"score": 0.46,
			"incomplete": false,
			"visualizationId": "lnsPie",
			"visualizationState": {
					"shape": "donut",
					"layers": [
							{
									"layerId": "ac2fc639-8958-40b8-a161-fdb474458be1",
									"primaryGroups": [
											"event.action"
									],
									"metrics": [
											"AVG(http.response.bytes)"
									],
									"numberDisplay": "percent",
									"categoryDisplay": "default",
									"legendDisplay": "default",
									"nestedLegend": false,
									"layerType": "data",
									"colorMapping": {
											"assignments": [],
											"specialAssignments": [
													{
															"rule": {
																	"type": "other"
															},
															"color": {
																	"type": "loop"
															},
															"touched": false
													}
											],
											"paletteId": "eui_amsterdam_color_blind",
											"colorMode": {
													"type": "categorical"
											}
									}
							}
					]
			},
			"keptLayerIds": [
					"ac2fc639-8958-40b8-a161-fdb474458be1"
			],
			"datasourceState": {
					"layers": {
							"ac2fc639-8958-40b8-a161-fdb474458be1": {
									"index": "26d07135e742ad73410e1328d956475126835c17f827570b86af22dfdee21f62",
									"query": {
											"esql": "from kbn-data-forge-fake_stack.admin-console-2* | STATS AVG(http.response.bytes) by event.action "
									},
									"columns": [
											{
													"columnId": "AVG(http.response.bytes)",
													"fieldName": "AVG(http.response.bytes)",
													"meta": {
															"type": "number"
													},
													"inMetricDimension": true
											},
											{
													"columnId": "event.action",
													"fieldName": "event.action",
													"meta": {
															"type": "string"
													}
											}
									],
									"timeField": "@timestamp"
							}
					},
					"indexPatternRefs": [
							{
									"id": "26d07135e742ad73410e1328d956475126835c17f827570b86af22dfdee21f62",
									"title": "kbn-data-forge-fake_stack.admin-console-2*",
									"timeField": "@timestamp"
							}
					]
			},
			"datasourceId": "textBased",
			"columns": 2,
			"changeType": "unchanged"
		}
	}									
}'
```

elasticmachine commented:
Pinging @elastic/obs-ux-management-team (Team:obs-ux-management)

mgiota commented on packages/kbn-investigation-shared/src/index.ts:
```diff
@@ -0,0 +1,9 @@
+ * Side Public License, v 1.
+ */
+
+export * from './rest_specs';
```
@kdelemme Ι see you export here only from rest_specs. Does it make sense to export also from schema folder here as well?

```
export * from './schema'
export * from './rest_specs';
```

mgiota commented on x-pack/plugins/observability_solution/investigate_app/public/pages/details/investigation_details_page.tsx:
```diff
@@ -7,7 +7,7 @@
 import { EuiButton, EuiButtonEmpty, EuiText } from '@elastic/eui';
 import { i18n } from '@kbn/i18n';
-import { alertOriginSchema } from '@kbn/investigation-shared';
+import { alertOriginSchema } from '@kbn/investigation-shared/src/schema';
```
How about you do following in `packages/kbn-investigation-shared/src/index.ts`

`export * from './schema';` 

And then do following in this file?
`import { alertOriginSchema } from '@kbn/investigation-shared/`

mgiota commented on x-pack/plugins/observability_solution/investigate_app/server/services/investigation_repository.ts:
```diff
@@ -6,10 +6,11 @@
 import { isLeft } from 'fp-ts/lib/Either';
-import { Investigation, StoredInvestigation, investigationSchema } from '../models/investigation';
-import { SO_INVESTIGATION_TYPE } from '../saved_objects/investigation';
+import { Investigation, StoredInvestigation } from '../models/investigation';
```
What is the difference between `Investigation` and `StoredInvestigation`?

kdelemme replied:
At the moment they are 100% identical, but when using more complex io-ts type, the serialized/deserialized type can be different.
For example, we could have a DateString type that handle DateString <-> Date type, and therefore when serialized (OutputOf), the type is a string (i.e. what we receive on the route, what we send back from the route, and how we store it), but within the backend services, we would use the Date type. Which is much more convenient that having to parse the string every time.

kdelemme replied:
Example: https://github.com/kdelemme/kibana/blob/main/x-pack/packages/kbn-slo-schema/src/schema/duration.ts#L13

mgiota commented on x-pack/plugins/observability_solution/investigate_app/server/models/investigation.ts:
```diff
@@ -5,22 +5,8 @@
  */
 
-import { alertOriginSchema, blankOriginSchema } from '@kbn/investigation-shared';
+import { investigationSchema } from '@kbn/investigation-shared/src/schema';
```
Can we import like this `import { investigationSchema } from '@kbn/investigation-shared';` and export `schema` folder directly in the index file of the kbn-investigation-shared package?

mgiota commented on packages/kbn-investigation-shared/src/rest_specs/create_item.ts:
```diff
@@ -0,0 +1,35 @@
+  path: t.type({
+    investigationId: t.string,
+  }),
+  body: investigationItemsSchema,
```
Is it here `investigationItemsSchema` or maybe `investigationItemSchema`

mgiota replied:
Oh right! It's either or, somehow I confused it and thought it is an array of all types of items. Makes total sense now

mgiota commented on packages/kbn-investigation-shared/src/rest_specs/create_item.ts:
```diff
@@ -0,0 +1,35 @@
+type CreateInvestigationItemInput = t.OutputOf<
+  typeof createInvestigationItemParamsSchema.props.body
+>;
+type CreateInvestigationItemParams = t.TypeOf<
```
I was trying to understand the difference between `CreateInvestigationItemParams` and `CreateInvestigationItemInput` and I didn't find `CreateInvestigationItemParams` being used anywhere. Can you explain the difference?

kdelemme replied:
I will remove the Params one since they are not used. 
Basically it's what I explained in another comment about the DateString <-> Date object.
OutputOf would be the type with the Date, while TypeOf would be the type with the String representation of the Date.

kdelemme commented:
@mgiota I've renamed all Input to be Params (so they are consistent with their related schema names, e.g. `somethingParamsSchema -> SomethingParams type`), and removed the unused type.

mgiota reviewed (approved):
LGTM!

mgiota commented on packages/kbn-investigation-shared/src/schema/investigation_item.ts:
```diff
@@ -0,0 +1,27 @@
+
+const investigationItemsSchema = esqlItemSchema; // replace with union with various item types
+
+const investigationItemSchema = t.intersection([
```
@kdelemme Can you explain me this type? I am aware of using `intersection` when mixing required and optional props. Here both props are required `t.type`, I didn't see any `t.partial` here. Or is intersection used here to support the union with various item types later on?

kdelemme replied:
Certainly. intersection is just a way to concatenate types to make a new one. In this case, I'm using intersection with some common props for all items, and then the item schema which will become a union of different item at some point.

mgiota commented on x-pack/plugins/observability_solution/investigate/common/types.ts:
```diff
@@ -36,13 +35,12 @@ export interface InvestigateWidget<
-  created: number;
-  last_updated: number;
+  createdAt: number;
+  createdBy: string;
```
@kdelemme Maybe we want to use here `createdBy: user.username` similar to what you use [here](https://github.com/elastic/kibana/pull/190852/files#diff-6d6b56d1cb8b912a74f6ee48feaff0633296fa0b9085d7757ca08aa5bdfd491aR50)?
